File: health_prompt/HealthDashboard/Views/HDWaterView.m
//
//  HDWaterView.m
//  HealthDashboard - 喝水波浪动画
//
//  Created by zhang, haizi on 2026/3/19.
//

#import "HDWaterView.h"

@interface HDWaterView ()
@property (nonatomic, assign) CGFloat level;
@property (nonatomic, strong) CAShapeLayer *waveLayer;
@property (nonatomic, strong) CADisplayLink *displayLink;
@property (nonatomic, assign) CGFloat waveOffset;
@end

@implementation HDWaterView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.clipsToBounds  = YES;
        self.layer.cornerRadius = 8;
        self.backgroundColor = [UIColor colorWithRed:0.10 green:0.18 blue:0.30 alpha:1.0];
        _level = 0.3;

        _waveLayer = [CAShapeLayer layer];
        _waveLayer.fillColor = [UIColor colorWithRed:0.20 green:0.60 blue:0.95 alpha:0.75].CGColor;
        [self.layer addSublayer:_waveLayer];

        // 波浪动画定时器
        _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(updateWave)];
        [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
    }
    return self;
}

- (void)updateWave {
    _waveOffset += 0.03;
    [self drawWave];
}

- (void)drawWave {
    CGFloat w = self.bounds.size.width;
    CGFloat h = self.bounds.size.height;
    CGFloat waterY = h * (1 - _level);
    CGFloat amplitude = 4.0;

    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(0, waterY)];
    for (CGFloat x = 0; x <= w; x++) {
        CGFloat y = waterY + amplitude * sin((x / w * 2 * M_PI) + _waveOffset);
        [path addLineToPoint:CGPointMake(x, y)];
    }
    [path addLineToPoint:CGPointMake(w, h)];
    [path addLineToPoint:CGPointMake(0, h)];
    [path closePath];
    _waveLayer.path = path.CGPath;
}

- (void)setWaterLevel:(CGFloat)level animated:(BOOL)animated {
    _level = MAX(0, MIN(1, level));
}

- (void)dealloc {
    [_displayLink invalidate];
}

@end

File: health_advanced_swift_spec/HealthDashboard/Views/HDWaterView.swift
//
//  HDWaterView.swift
//  HealthDashboard - 喝水波浪动画
//
//  Created by zhang, haizi on 2026/3/22.
//

import UIKit

/// 水波进度视图
@MainActor
class HDWaterView: UIView {

    // MARK: - Properties

    private var level: CGFloat = 0.3
    private let waveLayer = CAShapeLayer()
    private var displayLink: CADisplayLink?
    private var waveOffset: CGFloat = 0

    // MARK: - Init

    override init(frame: CGRect) {
        super.init(frame: frame)
        clipsToBounds = true
        layer.cornerRadius = 8
        backgroundColor = UIColor(red: 0.10, green: 0.18, blue: 0.30, alpha: 1)

        waveLayer.fillColor = UIColor(red: 0.20, green: 0.60, blue: 0.95, alpha: 0.75).cgColor
        layer.addSublayer(waveLayer)

        displayLink = CADisplayLink(target: self, selector: #selector(updateWave))
        displayLink?.add(to: .main, forMode: .common)
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    nonisolated deinit {
        // displayLink 会随对象释放自动失效
        // 在 willMove(toWindow:) 中处理生命周期
    }

    override func willMove(toWindow newWindow: UIWindow?) {
        super.willMove(toWindow: newWindow)
        if newWindow == nil {
            displayLink?.invalidate()
            displayLink = nil
        }
    }

    // MARK: - Wave Animation

    @objc private func updateWave() {
        waveOffset += 0.03
        drawWave()
    }

    private func drawWave() {
        let w = bounds.width
        let h = bounds.height
        let waterY = h * (1 - level)
        let amplitude: CGFloat = 4.0

        let path = UIBezierPath()
        path.move(to: CGPoint(x: 0, y: waterY))
        for x in stride(from: CGFloat(0), through: w, by: 1) {
            let y = waterY + amplitude * sin((x / w * 2 * .pi) + waveOffset)
            path.addLine(to: CGPoint(x: x, y: y))
        }
        path.addLine(to: CGPoint(x: w, y: h))
        path.addLine(to: CGPoint(x: 0, y: h))
        path.close()
        waveLayer.path = path.cgPath
    }

    // MARK: - Public

    func setWaterLevel(_ level: CGFloat, animated: Bool) {
        self.level = max(0, min(1, level))
    }
}
